### PR TITLE
WT-5927 Make checkpoint cursors work again

### PR DESCRIPTION
**Pull request by keithbostic**

### CODE REVIEW

**Comment by agorrod** [comment]
The one test failure in the most recent build was WT-8431

**Comment by agorrod** [comment]
I also added a number of additional variants to the patch build, to increase coverage. We don't need to wait for all of them to complete before merging if things look good.

File: src/btree/bt_handle.c [hunk 1]
```diff
@@ -551,12 +554,14 @@ __btree_conf(WT_SESSION_IMPL *session, WT_CKPT *ckpt)
...
+     * The only scenarios where the checkpoint run write generation number is less than the
+     * connection last checkpoint base write generation number are when rollback to stable doesn't
+     * happen during the recovery due to the unavailability of history store file, or when reading a
+     * checkpoint.
      */
-    if (!F_ISSET(conn, WT_CONN_RECOVERING) || F_ISSET(btree, WT_BTREE_LOGGED) ||
-      ckpt->run_write_gen < conn->last_ckpt_base_write_gen)
+    if ((!F_ISSET(conn, WT_CONN_RECOVERING) || F_ISSET(btree, WT_BTREE_LOGGED) ||
+          ckpt->run_write_gen < conn->last_ckpt_base_write_gen) &&
+      !is_ckpt)
```
Comment (kommiharibabu) [inline src/btree/bt_handle.c:564]: We shouldn't avoid setting the `base_write_gen` for unnamed checkpoints also. As it is possible that unnamed checkpoints don't write new checkpoints in the new session if there are no changes. So don't set it only for the named checkpoints.

**Comment by sauclovian-wt** [comment]
Not skipping trees on forced checkpoints has triggered a lot of fallout; still assessing the situation.

**Comment by sauclovian-wt** [comment]
Ok, the conclusion is: not skipping trees breaks most of the salvage tests (basic problem is: it opens all trees to checkpoint them, including ones that are supposed to need salvage, and in many cases they can't be opened, which leads to panics in the block manager; fixing this is a superset of WT-6431 and not trivial) and isn't going to be a viable approach anytime soon.

Instead I'm going to store the write generation from the global checkpoint metadata in the checkpoint cursor, substitute it into the session during checkpoint ops, and explicitly use it in preference to the tree's base write generation in the unpack cleanup code. If the global checkpoint is from a newer run than the tree checkpoint, which is possible, this forces us to not see the transaction IDs; otherwise it works as before.

Changes will arrive when Keith pushes them (since this is his branch and not mine so I can't push to it...)

Given that on several occasions in recent weeks there's been discussions with reasoning based on the (now demonstrably false) assumption that the recovery checkpoint doesn't skip trees, I'm a little worried that there are problems based on those assumptions waiting to crop up. So the other conclusion we came to when discussing this tonight was that we should probably turn off "force" on the recovery checkpoint, at least experimentally, and if anything breaks understand why.

File: src/btree/bt_handle.c (hunk 1 above)
Comment (sauclovian-wt) [reply src/btree/bt_handle.c:564]: (This comment is out of date, right?)

**Comment by keithbostic** [comment]
I've also schedule a full run here: https://evergreen.mongodb.com/version/625e07417742ae58d2da79f1?redirect_spruce_users=true

File: src/btree/bt_handle.c (hunk 1 above)
Comment (kommiharibabu) [reply src/btree/bt_handle.c:564]: Yes, Thanks.

**Comment by keithbostic** [comment]
Restarted full test run after rebase.

https://evergreen.mongodb.com/version/625ef79b3e8e8626a4e89f0c?redirect_spruce_users=true

File: src/include/cell_inline.h [hunk 2]
```diff
@@ -1082,8 +1084,23 @@ __cell_unpack_window_cleanup(WT_SESSION_IMPL *session, const WT_PAGE_HEADER *dsk
...
+        /*
+         * When reading a checkpoint, override the tree's base write generation with the write
+         * generation from the global metadata, which might be newer. This comes into play if the
+         * tree checkpoint is from an older database run than the global checkpoint, which can
+         * happen if checkpointing skips the tree at the right points. Bypass this logic if the
+         * checkpoint write generation isn't set because the checkpoint is from an older version of
+         * WiredTiger; in that case we use the tree's write generation and hope for the best.
+         */
+        write_gen = session->checkpoint_write_gen;
+        WT_ASSERT(session, write_gen >= S2BT(session)->base_write_gen);
```
Comment (kommiharibabu) [inline src/include/cell_inline.h:1085]: I have a doubt about this condition. In the following scenario,
1. connection base write gen is 10, so 10 is stored in the metadata as part of the checkpoint.
2. Restarted and connection base write gen is incremented to 20 and it is saved in the btree base write gen.
3. Reading it from the checkpoint can cause problems?

**Review by kommiharibabu — APPROVED** [review]
LGTM

File: src/include/cell_inline.h (hunk 2 above)
Comment (sauclovian-wt) [reply src/include/cell_inline.h:1085]: It isn't saved in the btree base write gen if it's a checkpoint. Recall this change from the original branch: [4570ff5](https://github.com/wiredtiger/wiredtiger/pull/7661/commits/4570ff593aa766c0533bf2ba6f972307bb495914); that makes the btree base write gen for checkpoint btrees always be the write gen from the tree-level checkpoint. The new problem last week was that the tree-level checkpoint can be older than the global checkpoint, and if so we want to use the global checkpoint's write generation instead. But this should only move the write generation forward; hence the assertion here.

Comment (kommiharibabu) [reply src/include/cell_inline.h:1085]: Thanks, I forgot the change in bt_handle.c.

**Pull request merged**